File: typescript/src/pool.ts
/**
 * Optional connection pool (CLT-080) — a layer **above** the
 * single-connection {@link Client} (CLT-001: "pooling is a layer above").
 *
 * Under a mandatory-`HELLO` config with `auth_required`, a fresh connection
 * costs a handshake round trip before the first request. A caller that opens a
 * connection per operation therefore pays that round trip every time. The pool
 * amortizes it: `N` operations over a checked-out connection pay **one**
 * connect and **one** handshake, not `N`.
 *
 * The shape is deliberately minimal — a fixed number of connections bounded by
 * a semaphore, an idle list, lazy connect on first checkout, and a guard that
 * returns the connection on {@link PooledConn.release}. It is **not** an
 * external pool library: health checks, background reaping and min-idle warmup
 * are out of scope; a poisoned connection (CLT-014) is dropped on release and
 * the next checkout connects fresh, leaving reconnect to CLT-030 rather than
 * the pool.
 *
 * The pool adds **no wire behavior**: it builds the same {@link Client} as
 * {@link Client.connect} from a {@link Config} and {@link ClientOptions}, and
 * the single-connection client's API is unchanged (CLT-001). `maxInFlight`
 * (CLT-012) stays a per-connection bound; the pool bounds connections, not
 * in-flight calls.
 *
 * TypeScript has no RAII drop, so the guard is explicit: call
 * {@link PooledConn.release} in a `finally`.
 *
 * ```ts
 * import { Config, Pool } from "@hivehub/thunder";
 *
 * const app = Config.standard().withScheme("myapp").withPort(9000);
 * const pool = new Pool("myapp://localhost", app, {}, 8);
 * const conn = await pool.acquire(); // reuses an idle connection, or dials one
 * try {
 *   const pong = await conn.client.call("PING");
 * } finally {
 *   conn.release(); // returns the connection to the pool for reuse
 * }
 * ```
 */

import { Client } from "./client";
import type { ClientOptions } from "./client";
import type { Config } from "./config";

/**
 * Bounds live + checked-out connections (CLT-080 "fixed N"). A checkout beyond
 * the limit awaits a release rather than opening another connection. FIFO, so
 * waiters proceed in order.
 */
class PoolSemaphore {
  #available: number;
  readonly #waiters: (() => void)[] = [];

  constructor(permits: number) {
    this.#available = permits;
  }

  acquire(): Promise<void> {
    if (this.#available > 0) {
      this.#available -= 1;
      return Promise.resolve();
    }
    return new Promise((resolve) => this.#waiters.push(resolve));
  }

  release(): void {
    const next = this.#waiters.shift();
    if (next) next();
    else this.#available += 1;
  }
}

/**
 * A bounded pool of {@link Client}s over one endpoint (CLT-080).
 *
 * At most `maxConnections` connections are live at once; a checkout beyond that
 * awaits a return. Connections are dialed lazily — construction opens none —
 * and reused across checkouts so the handshake is paid once per connection, not
 * once per operation.
 */
export class Pool {
  readonly #endpoint: string;
  readonly #config: Config;
  readonly #options: ClientOptions;
  readonly #permits: PoolSemaphore;
  /** Idle connections available for reuse. */
  readonly #idle: Client[] = [];

  /**
   * Build a pool for `endpoint`. Opens no connections — the first
   * {@link acquire} dials the first one. `maxConnections` is clamped to at
   * least 1.
   */
  constructor(
    endpoint: string,
    config: Config,
    options: ClientOptions = {},
    maxConnections = 1,
  ) {
    this.#endpoint = endpoint;
    this.#config = config;
    this.#options = options;
    this.#permits = new PoolSemaphore(Math.max(1, Math.floor(maxConnections)));
  }

  /**
   * Check out a connection. Reuses an idle, **live** connection when one is
   * available; otherwise dials and handshakes a fresh one (CLT-002). Awaits a
   * return when `maxConnections` are already checked out. The returned
   * {@link PooledConn} returns the connection to the pool on
   * {@link PooledConn.release}.
   */
  async acquire(): Promise<PooledConn> {
    await this.#permits.acquire();
    // Reuse the newest idle connection that is still live; discard any that
    // were poisoned (CLT-014) while sitting idle.
    let client: Client | null = null;
    for (;;) {
      const candidate = this.#idle.pop();
      if (candidate === undefined) break;
      if (candidate.isAlive) {
        client = candidate;
        break;
      }
      void candidate.close();
    }
    if (client === null) {
      try {
        client = await Client.connect(this.#endpoint, this.#config, this.#options);
      } catch (error) {
        // The dial failed: give the permit back so it is not leaked.
        this.#permits.release();
        throw error;
      }
    }
    return new PooledConn(client, this.#idle, this.#permits);
  }

  /**
   * Idle connections currently parked in the pool. For diagnostics and tests —
   * production code should not branch on it.
   */
  get idleCount(): number {
    return this.#idle.length;
  }
}

/**
 * Guard from {@link Pool.acquire}. Exposes the checked-out {@link Client} via
 * {@link client} and returns the connection to the pool on {@link release} so
 * the next checkout reuses it — unless the connection was poisoned, in which
 * case it is dropped and the next checkout connects fresh (CLT-014/030).
 *
 * TypeScript has no RAII drop, so {@link release} is explicit: call it in a
 * `finally`. It is idempotent — a second call is a no-op.
 */
export class PooledConn {
  #client: Client | null;
  readonly #idle: Client[];
  readonly #permits: PoolSemaphore;

  /** @internal — constructed only by {@link Pool.acquire}. */
  constructor(client: Client, idle: Client[], permits: PoolSemaphore) {
    this.#client = client;
    this.#idle = idle;
    this.#permits = permits;
  }

  /** The checked-out client. Throws after {@link release}. */
  get client(): Client {
    if (this.#client === null) {
      throw new Error("PooledConn used after release()");
    }
    return this.#client;
  }

  /**
   * Return the connection to the pool (idempotent). CLT-014: only a live
   * connection is parked for reuse; a poisoned or closed one is dropped here
   * and the next checkout dials fresh, leaving reconnect to CLT-030 rather
   * than the pool. Releasing the permit lets a waiting checkout proceed.
   */
  release(): void {
    const client = this.#client;
    if (client === null) return;
    this.#client = null;
    if (client.isAlive) {
      this.#idle.push(client);
    } else {
      void client.close();
    }
    this.#permits.release();
  }
}

File: rust/thunder/src/client/pool.rs
//! Optional connection pool (CLT-080) — a layer **above** the
//! single-connection [`Client`] (CLT-001: "pooling is a layer above").
//!
//! Under a mandatory-`HELLO` profile with `auth_required`, a fresh connection
//! costs a handshake round trip before the first request. A caller that opens a
//! connection per operation therefore pays that round trip every time — the
//! failure Nexus's per-request REST client and Vectorizer's per-RPC Raft
//! channel each hit as TIME_WAIT port exhaustion. The pool amortizes it: `N`
//! operations over a checked-out connection pay **one** connect and **one**
//! handshake, not `N`.
//!
//! The shape is deliberately minimal — a fixed number of connections bounded by
//! a semaphore, an idle list, lazy connect on first checkout, and an RAII guard
//! that returns the connection on drop. It is **not** `bb8`/`deadpool`/`r2d2`:
//! those bring async traits and heavyweight reconnect logic this layer does not
//! need. Health checks, background reaping and min-idle warmup are out of scope;
//! a poisoned connection (CLT-014) is dropped on return and the next checkout
//! connects fresh, leaving reconnect to CLT-030 rather than the pool.
//!
//! The pool adds **no wire behavior**: it builds the same [`Client`] as
//! [`Client::connect_with`] from a [`Config`] and [`ClientConfig`], and the
//! single-connection client's API is unchanged (CLT-001). `max_in_flight`
//! (CLT-012) stays a per-connection bound; the pool bounds connections, not
//! in-flight calls.
//!
//! ```no_run
//! use thunder::{ClientConfig, Config};
//! use thunder::client::Pool;
//!
//! # async fn demo() -> Result<(), thunder::ClientError> {
//! let app = Config::standard().scheme("myapp").port(9000);
//! let pool = Pool::new("myapp://localhost", app, ClientConfig::new(), 8);
//! let conn = pool.acquire().await?; // reuses an idle connection, or dials one
//! let pong = conn.call("PING", vec![]).await?;
//! assert_eq!(pong.as_str(), Some("PONG"));
//! // `conn` returns the connection to the pool when it drops.
//! # Ok(())
//! # }
//! ```

use std::sync::{Arc, Mutex as StdMutex, MutexGuard, PoisonError};

use tokio::sync::{OwnedSemaphorePermit, Semaphore};

use crate::client::{Client, ClientConfig, ClientError};
use crate::wire::Config;

/// Ride through std-mutex poisoning: a panicked holder must not wedge the pool.
fn lock<T>(mutex: &StdMutex<T>) -> MutexGuard<'_, T> {
    mutex.lock().unwrap_or_else(PoisonError::into_inner)
}

/// A bounded pool of [`Client`]s over one endpoint (CLT-080).
///
/// At most `max_connections` connections are live at once; a checkout beyond
/// that awaits a return. Connections are dialed lazily — construction opens
/// none — and reused across checkouts so the handshake is paid once per
/// connection, not once per operation.
pub struct Pool {
    endpoint: String,
    config: Config,
    client_config: ClientConfig,
    /// Bounds live + checked-out connections to `max_connections`.
    permits: Arc<Semaphore>,
    /// Idle connections available for reuse.
    idle: Arc<StdMutex<Vec<Client>>>,
}

impl Pool {
    /// Build a pool for `endpoint`. Opens no connections — the first
    /// [`acquire`](Self::acquire) dials the first one. `max_connections` is
    /// clamped to at least 1.
    pub fn new(
        endpoint: impl Into<String>,
        config: Config,
        client_config: ClientConfig,
        max_connections: usize,
    ) -> Self {
        let max = max_connections.max(1);
        Self {
            endpoint: endpoint.into(),
            config,
            client_config,
            permits: Arc::new(Semaphore::new(max)),
            idle: Arc::new(StdMutex::new(Vec::with_capacity(max))),
        }
    }

    /// Check out a connection. Reuses an idle, **live** connection when one is
    /// available; otherwise dials and handshakes a fresh one (CLT-002). Awaits a
    /// return when `max_connections` are already checked out. The returned
    /// [`PooledConn`] returns the connection to the pool on drop.
    pub async fn acquire(&self) -> Result<PooledConn, ClientError> {
        let permit = Arc::clone(&self.permits)
            .acquire_owned()
            .await
            .map_err(|_| ClientError::Connection {
                message: "connection pool is closed".to_owned(),
            })?;

        // Reuse the newest idle connection that is still live; discard any that
        // were poisoned (CLT-014) while sitting idle.
        let reused = {
            let mut idle = lock(&self.idle);
            loop {
                match idle.pop() {
                    Some(client) if client.is_alive() => break Some(client),
                    Some(_dead) => continue,
                    None => break None,
                }
            }
        };
        let client = match reused {
            Some(client) => client,
            None => {
                Client::connect_with(
                    &self.endpoint,
                    self.config.clone(),
                    self.client_config.clone(),
                )
                .await?
            }
        };

        Ok(PooledConn {
            inner: Some(client),
            idle: Arc::clone(&self.idle),
            _permit: permit,
        })
    }

    /// Idle connections currently parked in the pool. For diagnostics and tests
    /// — production code should not branch on it.
    pub fn idle_count(&self) -> usize {
        lock(&self.idle).len()
    }
}

/// RAII guard from [`Pool::acquire`]. Derefs to the [`Client`], and returns the
/// connection to the pool on drop so the next checkout reuses it — unless the
/// connection was poisoned, in which case it is dropped and the next checkout
/// connects fresh (CLT-014/030).
pub struct PooledConn {
    /// `Some` for the guard's whole life; taken only in [`Drop`].
    inner: Option<Client>,
    idle: Arc<StdMutex<Vec<Client>>>,
    /// Held for the checkout's duration; releasing it lets a waiter proceed.
    _permit: OwnedSemaphorePermit,
}

impl PooledConn {
    /// Borrow the checked-out client. (Also available via [`Deref`].)
    pub fn client(&self) -> &Client {
        match &self.inner {
            Some(client) => client,
            // Unreachable: `inner` is only taken in `Drop`, after which no
            // method can be called on the guard.
            None => unreachable!("PooledConn::client after drop"),
        }
    }
}

impl std::ops::Deref for PooledConn {
    type Target = Client;

    fn deref(&self) -> &Client {
        self.client()
    }
}

impl Drop for PooledConn {
    fn drop(&mut self) {
        if let Some(client) = self.inner.take() {
            // CLT-014: only a live connection returns to the pool. A poisoned or
            // closed one is dropped here; the next checkout dials fresh, leaving
            // reconnect to CLT-030 rather than the pool.
            if client.is_alive() {
                lock(&self.idle).push(client);
            }
        }
    }
}

#[cfg(test)]
#[allow(clippy::unwrap_used, clippy::expect_used)]
mod tests {
    use super::*;

    #[test]
    fn new_does_not_dial_and_clamps_capacity() {
        let pool = Pool::new(
            "test://127.0.0.1:0",
            Config::standard(),
            ClientConfig::new(),
            0,
        );
        // No connection opened at construction, and max clamped to >= 1.
        assert_eq!(pool.idle_count(), 0);
        assert_eq!(pool.permits.available_permits(), 1);
    }
}
